internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1120 @@ package controller
+		if existingCsiDriver.UID != "" {
+			log.Info("CSI Driver resource exist but does not meet desired state")
+			if err := r.Delete(r.ctx, existingCsiDriver); err != nil {
+				log.Error(err, "Failed to delete existing CSI Driver resource")
+				return err
+			}
+			log.Info("CSI Driver resource deleted successfully")
+		} else {
+			log.Info("CSI Driver resource does not exist")
+		}
+
+		if err := r.Create(r.ctx, desiredCsiDriver); err != nil {
+			log.Error(err, "Failed to create a CSI Driver resource")
+			return err
+		}
+
+		log.Info("CSI Driver resource created successfully")
+	} else {
+		log.Info("CSI Driver resource already meets desired state")
+	}
+
+	return nil
+}
+
+func (r *driverReconcile) reconcileControllerPluginDeployment() error {
+	deploy := &appsv1.Deployment{}
+	deploy.Name = r.generateName("ctrlplugin")
+	deploy.Namespace = r.driver.Namespace
+
+	log := r.log.WithValues("deploymentName", deploy.Name)
+	log.Info("Reconciling controller plugin deployment")
+
+	opResult, err := ctrlutil.CreateOrUpdate(r.ctx, r.Client, deploy, func() error {
+		r.log.Info("Controller plugin deployment reconciled successfully")
+		if err := ctrlutil.SetOwnerReference(&r.driver, deploy, r.Scheme); err != nil {
+			log.Error(err, "Failed setting an owner reference on deployment")
+			return err
+		}
+
+		appName := deploy.Name
+		leaderElectionSpec := utils.FirstNonNil(r.driver.Spec.LeaderElection, &defaultLeaderElection)
+		pluginSpec := utils.FirstNonNil(r.driver.Spec.ControllerPlugin, &csiv1a1.ControllerPluginSpec{})
+		serviceAccountName := utils.FirstNonEmpty(
+			ptr.Deref(pluginSpec.ServiceAccountName, ""),
+			fmt.Sprintf("csi-%s-ctrlplugin-sa", r.driverType),
+		)
+		imagePullPolicy := utils.FirstNonEmpty(pluginSpec.ImagePullPolicy, corev1.PullIfNotPresent)
+		grpcTimeout := utils.FirstNonZero(r.driver.Spec.GRpcTimeout, defaultGRrpcTimeout)
+		logLevel := utils.If(r.driver.Spec.Log != nil, r.driver.Spec.Log.LogLevel, 0)
+		forceKernelClient := r.isCephFsDriver() && r.driver.Spec.CephFsClientType == csiv1a1.KernelCephFsClient
+
+		leaderElectionArgs := []string{
+			utils.LeaderElectionContainerArg,
+			utils.LeaderElectionNamespaceContainerArg(r.driver.Namespace),
+			utils.LeaderElectionLeaseDurationContainerArg(leaderElectionSpec.LeaseDuration),
+			utils.LeaderElectionRenewDeadlineContainerArg(leaderElectionSpec.RenewDeadline),
+			utils.LeaderElectionRetryPeriodContainerArg(leaderElectionSpec.RetryPeriod),
+		}
+
+		deploy.Spec = appsv1.DeploymentSpec{
+			Replicas: pluginSpec.Replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{"app": appName},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: utils.Call(func() map[string]string {
+						podLabels := map[string]string{}
+						maps.Copy(podLabels, pluginSpec.Labels)
+						podLabels["app"] = appName
+						return podLabels
```

Comment:
am I understanding it correctly, you are ensuring that deployment always own the pod w/ `app: <appName` irrespective and so overriding `podLabels["app"]` if it was defined in the driver spec that doesn't match `selector.matchLabels`?

I think this is good, unless I'm misunderstanding.

Reply:
Yes you do, I treat the `app` label as a reserved label, and does not allow overwriting it

Reply:
@nb-ohad could you pls mention this in API comments? Controllers using new API should know `app` label is reserved.

We do have a side effect wrt change in names, client/rook might be deploying CSI with name csi-xyz and this operator translates that to csi-abc and those operators need to cleanup old deployments during the transition.